data_steward/cdr_cleaner/cleaning_rules/calculate_bmi.py

```diff
@@ -0,0 +1,187 @@
+"""
+Some participants self-report their height and weight. This CR calculates the BMI for
+those self-reported measurements. (Note BMIs are already calculated if the data is from HPO sites.)
+
+Each participant can self-report their height and weight multiple times. BMIs are calculated
+for each self-reported hight/weight pair. BMIs are not calculated if either height or weight is missing.
+
+Original Issue: DC-3239
+"""
+# Python imports
+import logging
+
+# Project imports
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+from cdr_cleaner.cleaning_rules.clean_height_weight import CleanHeightAndWeight
+from common import JINJA_ENV, MEASUREMENT
+from constants.cdr_cleaner.clean_cdr import (CONTROLLED_TIER_DEID_CLEAN, QUERY,
+                                             REGISTERED_TIER_DEID_CLEAN)
+from utils import pipeline_logging
+
+LOGGER = logging.getLogger(__name__)
+
+CREATE_SANDBOX_QUERY = JINJA_ENV.from_string("""
+CREATE OR REPLACE TABLE `{{project}}.{{sandbox_dataset}}.{{sandbox_table}}` AS (
+WITH self_reported_height AS (
+    SELECT
+        m.measurement_id, 
+        m.person_id, 
+        m.visit_occurrence_id, 
+        m.value_as_number,
+        m.measurement_date,
+        m.measurement_datetime,
+        m.measurement_time
+    FROM `{{project}}.{{dataset}}.measurement` m
+    JOIN `{{project}}.{{dataset}}.measurement_ext` e
+    ON m.measurement_id = e.measurement_id
+    WHERE e.src_id LIKE 'Participant Portal%'
+    AND m.measurement_concept_id = 3036277
+), self_reported_weight AS (
+    SELECT
+        m.measurement_id, 
+        m.person_id, 
+        m.visit_occurrence_id, 
+        m.value_as_number
+    FROM `{{project}}.{{dataset}}.measurement` m
+    JOIN `{{project}}.{{dataset}}.measurement_ext` e
+    ON m.measurement_id = e.measurement_id
+    WHERE e.src_id LIKE 'Participant Portal%'
+    AND m.measurement_concept_id = 3025315
+)
+SELECT
+    ROW_NUMBER() OVER(
+        ORDER BY h.person_id, h.visit_occurrence_id
+    ) + (
+        SELECT MAX(measurement_id) 
+        FROM `{{project}}.{{dataset}}.measurement`
+    ) AS measurement_id,
+    h.person_id,
+    3038553 AS measurement_concept_id,
+    h.measurement_date,
+    h.measurement_datetime,
+    h.measurement_time,
+    32865 AS measurement_type_concept_id,
+    NULL AS operator_concept_id,
+    w.value_as_number / POW(h.value_as_number / 100, 2) AS value_as_number,
+    NULL AS value_as_concept_id,
+    9531 AS unit_concept_id,
+    NULL AS range_low,
+    NULL AS range_high,
+    NULL AS provider_id,
+    h.visit_occurrence_id,
+    NULL AS visit_detail_id,
+    CAST(NULL AS STRING) AS measurement_source_value,
+    NULL AS measurement_source_concept_id,
+    'kg/m2' AS unit_source_value,
+    CAST(NULL AS STRING) AS value_source_value
+FROM self_reported_height h
+JOIN self_reported_weight w
+ON h.person_id = w.person_id AND h.visit_occurrence_id = w.visit_occurrence_id
```

Comment:
@hiro-mishima why don't you add probably measurment_date as a condition to match here? It is not required when we are matching on visit_occurrence_id but I believe having measurement_date will add an extra layer of conformance in case there is any issue with the visit_occurrence_ids that may be duplicated or what not.

Reply:
Thanks! I added `measurement_date` as a JOIN condition. I also added `src_id` to the condition just to be safer.